nsm/scripts/generate-routes.js

```diff
@@ -5,7 +5,146 @@ const prettier = require("prettier")
 const { existsSync } = require("fs")
 const fs = require("fs/promises")
 
-async function generateRoutes() {
+const VALID_EXTENSIONS = [".tsx", ".jsx", ".ts", ".js"]
+
+async function generateNsmPagesManifest(rootDir, outputDir) {
+  const pagesDir = existsSync(path.resolve(rootDir, "pages"))
+    ? path.resolve(rootDir, "pages")
+    : path.resolve(rootDir, "src/pages")
+  const pageFiles = glob.sync("**/*.{ts,tsx,js,jsx,cjs,mjs,mjsx,cjsx}", { cwd: pagesDir })
+
+  const manifest = {}
+
+  for (const pageFile of pageFiles) {
+    const relativePagePath = path.relative(process.cwd(), pageFile)
+    const parsedPagePath = path.parse(relativePagePath)
+
+    const isInvalidExtension = !VALID_EXTENSIONS.includes(
+      parsedPagePath.ext
+    )
+
+    if (isInvalidExtension) {
+      throw new Error(
+        `Invalid file extension for page: ${pageFile}, must be one of: ${VALID_EXTENSIONS.join(', ')}. Found: ${pageName.ext}`
+      )
+    }
+
+    const filenameWithoutExtension = parsedPagePath.base
+      .replace(parsedPagePath.ext, "")
+      .replace("index", "")
+
+    const filePathWithoutExtension = path.join(
+      parsedPagePath.dir,
+      filenameWithoutExtension
+    )
+
+    manifest[`/${filePathWithoutExtension}`] = `pages/${relativePagePath}`
+  }
+
+  const manifestJSON = JSON.stringify(manifest, null, 2)
+
+  const manifestFilePath = path.join(outputDir, "pages-manifest.json")
+
+  // there is no need to write the manifest file
+  // but it's good to have it for debugging purposes
+  await fs.writeFile(manifestFilePath, manifestJSON)
+
+  console.log(`Pages manifest generated at: ${manifestFilePath}`)
+
+  return manifest
+}
+
+function generateRouteFile({
+  pagesDirRelativePath,
+  pagesManifest,
+  nextless,
+  onlyApiFiles,
+  staticFiles,
+}) {
+  const staticFilesString = nextless
+    ? ""
+    : `,
+  ${staticFiles
+    .filter((fp) => fp.startsWith("static/"))
+    .map(
+      (fp) =>
+        `"/_next/${fp}": serveStatic("${
+          fp.split(".").slice(-1)[0]
+        }", require("./generated_static/${fp}.ts").default)`
+    )}`
+
+  return prettier.format(
+    `// @ts-nocheck
+    import serveStatic from "./serve-static"
+    export default {
+  ${Object.entries(pagesManifest)
+    .filter(([route, fp]) => {
+      if (!nextless) {
+        return true
+      }
+
+      if (!onlyApiFiles && !fp.startsWith("pages/api")) {
+        throw new Error(
+          `At the moment, only API routes are supported. Found: ${fp}`
+        )
+      }
+
+      if (
+        (onlyApiFiles && !fp.startsWith("pages/api")) ||
+        (!nextless && onlyApiFiles && !fp.startsWith("pages/api"))
+      ) {
+        return false
+      }
+
+      return true
+    })
+    .map(([route, fp]) => {
+      const fpNoExt = fp.split(".").slice(0, -1).join(".")
```

Comment:
Here and in other places, I think we can do what we need with methods from 'node:path' instead of direct string manipulation. It should be safer and more clear.

Reply:
Let's create another issue to solve this to not block this PR